README.md
# forwardcache

[![Build Status](https://travis-ci.org/mikegleasonjr/forwardcache.svg?branch=master)](https://travis-ci.org/mikegleasonjr/forwardcache)
[![Coverage Status](https://codecov.io/gh/mikegleasonjr/forwardcache/branch/master/graph/badge.svg)](https://codecov.io/gh/mikegleasonjr/forwardcache)
[![GitHub license](https://img.shields.io/badge/license-Apache%202-blue.svg)](https://raw.githubusercontent.com/mikegleasonjr/forwardcache/master/LICENSE)
[![GoDoc](http://godoc.org/github.com/mikegleasonjr/forwardcache?status.svg)](http://godoc.org/github.com/mikegleasonjr/forwardcache)

A distributed forward caching proxy for Go's http.Client using [httpcache][httpcache] and heavily inspired by [groupcache][groupcache]. Backed by a lot of existing cache [backends][backends] thanks to httpcache. A per host LRU algorithm is provided to optionally front any existing cache. Like groupcache, forwardcache "is a client library as well as a server. It connects to its own peers."

In simple terms, it is a distributed cache for HEAD and GET requests. It follows the HTTP RFC so it will only cache cacheable responses (like browsers do).

Docs on [godoc.org][godoc]

<pre>
+---------------------------------+
|                                 |
|         origin servers          |
|                                 |
+--+-------+-------+-----------+--+
   ^       ^       ^           ^
   |       |       |           |
   |       |       |           |
+-----+ +-----+ +-----+     +-----+
|cache| |cache| |cache|     |cache|
+--+--+ +--+--+ +--+--+     +--+--+
|     | |     | |     |     |     |
|  1  +-+  2  +-+  3  +--|--+  N  |
|     | |     | |     |     |     |
+-----+ +-----+ +-----+     +-----+
</pre>

## Requirements

* Go 1.7 (using request's context)

## Motivation

* Needed requests to be cached
* Needed bandwidth to be spread among N peers
* Needed to cache HTTPS requests
* Needed a lightweight infrastructure

So I couldn't go with existing solutions like Apache Traffic Server or Squid.

## Process

When making a request through the pool...

1. the requested url is hashed to determine which peer is responisble to proxy the request to the origin
2. the peer is called, if the request is cached and valid, it is returned without contacting the origin
3. if the request is not cached, the request is fetched from the origin and cached if cacheable before being returned back to the client

## Example

```go
peer := NewPeer("http://10.0.1.1:3000")
peer.SetPool("http://10.0.1.1:3000", "http://10.0.1.2:3000")

// -or-

client := NewClient(WithPool("http://10.0.1.1:3000", "http://10.0.1.2:3000"))
peer = NewPeer("http://10.0.1.1:3000", WithClient(client))

// -then-

http.DefaultTransport = peer
http.Get("https://...js/1.5.7/angular.min.js")

// -or-

http.DefaultClient = peer.HTTPClient()
http.Get("https://...js/1.5.7/angular.min.js")

// -or-

c := peer.HTTPClient()
c.Get("https://...js/1.5.7/angular.min.js")

// ...

http.ListenAndServe(":3000", peer.Handler())
```

## Licence

Apache 2.0 (see LICENSE and NOTICE)






[httpcache]: https://github.com/gregjones/httpcache  "gregjones/httpcache"
[backends]: https://github.com/gregjones/httpcache#cache-backends  "cache backends"
[groupcache]: https://github.com/gregjones/httpcache#cache-backends  "golang/groupcache"
[singleflight]: https://godoc.org/golang.org/x/sync/singleflight "x/sync/singleflight"
[godoc]: http://godoc.org/github.com/mikegleasonjr/forwardcache "mikegleasonjr/forwardcache"


CHANGELOG.md
## Change Log

v2.0.0 - 20/10/2016

* Cleaner API
* Performance improvements (`http.BufferPool`)
* Faster tests (using mocked transports instead of http servers)

v1.0.1 - 20/10/2016

* Can now specify custom `http.Transport`s
* We are now using functional options
  * see [http://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis][fo]
* API breaking change: `client.Client` is now `client.HTTPClient` (sorry I know I bumped to 1.0.0 yesterday)

v1.0.0 - 19/10/2016

* Can now use a nonparticipating client (a client that is not a peer in the pool)



[fo]: http://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis  "gregjones/httpcache"


peer_test.go
/*
Copyright 2018 Mike Gleason jr Couturier.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package forwardcache

import (
	"bytes"
	"net/http"
	"testing"
)

func TestPeer(t *testing.T) {
	hash := newHashMock().
		with("http://self.com:3000", 0).
		with("http://peer.com:3000", 1).
		with("http://some.url/res-self.js", 0).
		with("http://some.url/res-peer.js", 1)

	peerTransport := roundTripperFunc(func(*http.Request) (*http.Response, error) {
		res := okResponse()
		res.Header.Set("X-Source", "self")
		return res, nil
	})

	clientTransport := roundTripperFunc(func(req *http.Request) (*http.Response, error) {
		res := okResponse()
		res.Header.Set("X-Source", "client")
		return res, nil
	})

	peer := NewPeer("http://self.com:3000",
		WithPeerTransport(peerTransport),
		WithClient(NewClient(
			WithPool("http://self.com:3000", "http://peer.com:3000"),
			WithHashFn(hash.fn),
			WithClientTransport(clientTransport),
		)),
	)

	testCases := []struct {
		url  string
		want string
	}{
		{"http://some.url/res-self.js", "self"},
		{"http://some.url/res-peer.js", "client"},
	}
	for _, tC := range testCases {
		t.Run(tC.url, func(t *testing.T) {
			req, _ := http.NewRequest("GET", tC.url, nil)
			res, err := peer.RoundTrip(req)
			// io.Copy(ioutil.Discard, res.Body)
			// res.Body.Close()

			if err != nil {
				t.Fatalf("unexpected error: got %q, want <nil>", err)
			}

			if got := res.Header.Get("X-Source"); got != tC.want {
				t.Fatalf("unexpected transport handler: got %q, want %q", got, tC.want)
			}
		})
	}
}

func ExampleNewPeer() {
	peer := NewPeer("http://10.0.1.1:3000")
	peer.SetPool("http://10.0.1.1:3000", "http://10.0.1.2:3000")

	// -or-

	client := NewClient(WithPool("http://10.0.1.1:3000", "http://10.0.1.2:3000"))
	peer = NewPeer("http://10.0.1.1:3000", WithClient(client))

	// -then-

	http.DefaultTransport = peer
	http.Get("https://...js/1.5.7/angular.min.js")

	// -or-

	http.DefaultClient = peer.HTTPClient()
	http.Get("https://...js/1.5.7/angular.min.js")

	// -or-

	c := peer.HTTPClient()
	c.Get("https://...js/1.5.7/angular.min.js")

	// ...

	http.ListenAndServe(":3000", peer.Handler())
}

type hashMock struct {
	mocks  map[string]uint32
	hashTo uint32
}

func newHashMock() *hashMock { return &hashMock{mocks: map[string]uint32{}} }

func (h *hashMock) with(url string, hashTo uint32) *hashMock {
	h.mocks[url] = hashTo
	return h
}

func (h *hashMock) fn(data []byte) uint32 {
	for k, v := range h.mocks {
		if bytes.Contains(data, []byte(k)) {
			return v
		}
	}
	return h.hashTo
}


proxy_test.go
/*
Copyright 2018 Mike Gleason jr Couturier.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package forwardcache

import (
	"io"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"testing"
	"time"

	"github.com/gregjones/httpcache"
)

func TestProxy(t *testing.T) {
	okRoundTrip := roundTripperFunc(func(req *http.Request) (*http.Response, error) {
		res := okResponse()
		res.Header.Set("X-Url", req.URL.String())
		return res, nil
	})

	origin := newRoundTripperMock().
		add("GET", "http://cdn.com/jquery.js", okRoundTrip).
		add("POST", "http://cdn.com/jquery.js", okRoundTrip)

	proxy := newProxy("/p", httpcache.NewMemoryCache(), origin, DefaultBufferPool)

	testCases := []struct {
		method     string
		path       string
		status     int
		body       string
		xFromCache string
		xURL       string
	}{
		{"GET", "/another", http.StatusBadGateway, "", "", ""},
		{"GET", "/p?q=", http.StatusBadGateway, "", "", ""},
		{"GET", "/p?q=" + url.QueryEscape("http://10.0.1.%31/"), http.StatusBadGateway, "", "", ""},
		{"GET", "/p?q=" + url.QueryEscape("http://cdn.com/jquery.js"), http.StatusOK, "OK", "", "http://cdn.com/jquery.js"},
		{"GET", "/p?q=" + url.QueryEscape("http://cdn.com/jquery.js"), http.StatusOK, "OK", "1", "http://cdn.com/jquery.js"},
		{"POST", "/p?q=" + url.QueryEscape("http://cdn.com/jquery.js"), http.StatusOK, "OK", "", "http://cdn.com/jquery.js"},
		{"POST", "/p?q=" + url.QueryEscape("http://cdn.com/jquery.js"), http.StatusOK, "OK", "", "http://cdn.com/jquery.js"},
		{"GET", "/p?q=" + url.QueryEscape("http://cdn.com/bootstrap.js"), http.StatusNotFound, "Not Found", "", ""},
		{"GET", "/p?q=" + url.QueryEscape("http://cdn.com/bootstrap.js"), http.StatusNotFound, "Not Found", "", ""},
	}
	for _, tC := range testCases {
		path, _ := url.QueryUnescape(tC.path)
		t.Run(tC.method+path, func(t *testing.T) {
			rr := httptest.NewRecorder()
			req, _ := http.NewRequest(tC.method, tC.path, nil)
			proxy.ServeHTTP(rr, req)

			if rr.Code != tC.status {
				t.Errorf("proxy sent wrong status: got %d, want %d", rr.Code, tC.status)
			}

			if body := rr.Body.String(); body != tC.body {
				t.Errorf("proxy returned unexpected body: got %s want %s", body, tC.body)
			}

			if xFromCache := rr.HeaderMap.Get(httpcache.XFromCache); xFromCache != tC.xFromCache {
				t.Errorf("unexpected %q header: got %q, want %q", httpcache.XFromCache, xFromCache, tC.xFromCache)
			}

			if xURL := rr.HeaderMap.Get("X-Url"); xURL != tC.xURL {
				t.Errorf("unexpected X-Url header: got %q, want %q", xURL, tC.xURL)
			}
		})
	}
}

func BenchmarkProxy(b *testing.B) {
	body := strings.NewReader("OK")
	res := okResponse()
	origin := roundTripperFunc(func(req *http.Request) (*http.Response, error) {
		body.Seek(0, io.SeekStart)
		res.Body = ioutil.NopCloser(body)
		res.Request = req
		return res, nil
	})

	handler := newProxy("/proxy", &noopCache{}, origin, DefaultBufferPool)
	discard := &discarder{}
	req, _ := http.NewRequest("GET", "/proxy?q="+url.QueryEscape("http://cdn.com/jquery.js"), nil)

	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		handler.ServeHTTP(discard, req)
	}
}

type discarder struct{}

func (*discarder) Header() http.Header         { return make(http.Header) }
func (*discarder) Write(b []byte) (int, error) { return len(b), nil }
func (*discarder) WriteHeader(int)             {}

type noopCache struct{}

func (c *noopCache) Set(key string, resp []byte)   {}
func (c *noopCache) Delete(key string)             {}
func (c *noopCache) Get(key string) ([]byte, bool) { return nil, false }

type roundTripperFunc func(*http.Request) (*http.Response, error)

func (rt roundTripperFunc) RoundTrip(req *http.Request) (*http.Response, error) { return rt(req) }

type roundTripperMock struct {
	mocks map[string]roundTripperFunc
}

func newRoundTripperMock() *roundTripperMock {
	return &roundTripperMock{map[string]roundTripperFunc{}}
}

func (m *roundTripperMock) add(method, url string, f roundTripperFunc) *roundTripperMock {
	m.mocks[method+" "+url] = f
	return m
}

func (m *roundTripperMock) RoundTrip(req *http.Request) (*http.Response, error) {
	if f, ok := m.mocks[req.Method+" "+req.URL.String()]; ok {
		return f(req)
	}

	return &http.Response{
		Status:        "404 Not Found",
		StatusCode:    http.StatusNotFound,
		Header:        make(http.Header),
		Body:          ioutil.NopCloser(strings.NewReader("Not Found")),
		ContentLength: 9,
		Request:       req,
	}, nil
}

func okResponse() *http.Response {
	return &http.Response{
		StatusCode:    http.StatusOK,
		Body:          ioutil.NopCloser(strings.NewReader("OK")),
		ContentLength: 2,
		Header: map[string][]string{
			"date":    []string{time.Now().Format(time.RFC1123)},
			"Expires": []string{time.Now().Add(time.Hour).Format(time.RFC1123)},
		},
	}
}


client_test.go
package forwardcache

import (
	"io"
	"io/ioutil"
	"net/http"
	"net/url"
	"strings"
	"testing"
)

func TestClient(t *testing.T) {
	hash := newHashMock().
		with("http://a.com:3000", 0).
		with("http://b.com:3000", 1).
		with("http://c.com:3000", 2).
		with("http://some.url/res-a.js", 0).
		with("http://some.url/res-b.js", 1).
		with("http://some.url/res-c.js", 2)

	transport := roundTripperFunc(func(req *http.Request) (*http.Response, error) {
		res := okResponse()
		res.Header.Set("X-Requested-URL", req.URL.String())
		return res, nil
	})

	client := NewClient(
		WithPool("http://a.com:3000", "http://b.com:3000", "http://c.com:3000"),
		WithHashFn(hash.fn),
		WithClientTransport(transport),
		WithPath("/p"),
	).HTTPClient()

	testCases := []struct {
		url  string
		want string
	}{
		{"http://some.url/res-a.js", "http://a.com:3000/p?q=" + url.QueryEscape("http://some.url/res-a.js")},
		{"http://some.url/res-b.js", "http://b.com:3000/p?q=" + url.QueryEscape("http://some.url/res-b.js")},
		{"http://some.url/res-c.js", "http://c.com:3000/p?q=" + url.QueryEscape("http://some.url/res-c.js")},
	}
	for _, tC := range testCases {
		t.Run(tC.url, func(t *testing.T) {
			res, err := client.Get(tC.url)
			io.Copy(ioutil.Discard, res.Body)
			res.Body.Close()

			if err != nil {
				t.Fatalf("unexpected error: got %q, want <nil>", err)
			}

			if got := res.Header.Get("X-Requested-URL"); got != tC.want {
				t.Fatalf("malformed request to peer: got %q, want %q", got, tC.want)
			}
		})
	}
}

func ExampleNewClient() {
	client := NewClient(WithPool("http://10.0.1.1:3000", "http://10.0.1.2:3000"))

	// -then-

	http.DefaultTransport = client
	http.Get("https://...js/1.5.7/angular.min.js")

	// -or-

	http.DefaultClient = client.HTTPClient()
	http.Get("https://...js/1.5.7/angular.min.js")

	// -or-

	c := client.HTTPClient()
	c.Get("https://...js/1.5.7/angular.min.js")
}

func BenchmarkClient(b *testing.B) {
	body := strings.NewReader("OK")
	res := okResponse()
	peer := roundTripperFunc(func(req *http.Request) (*http.Response, error) {
		body.Seek(0, io.SeekStart)
		res.Body = ioutil.NopCloser(body)
		res.Request = req
		return res, nil
	})

	client := NewClient(
		WithClientTransport(peer),
		WithPool("http://localhost"),
	).HTTPClient()

	req, _ := http.NewRequest("GET", "http://cdn.com/jquery.js", nil)
	buff := make([]byte, 32*1024)

	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		res, _ := client.Do(req)
		io.CopyBuffer(ioutil.Discard, res.Body, buff)
		res.Body.Close()
	}
}


peer.go
/*
Copyright 2018 Mike Gleason jr Couturier.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package forwardcache provides a forward caching proxy that works
// across a set of peer processes. In simple terms, it is a distributed
// cache for HEAD and GET requests. It follows the HTTP RFC so it will
// only cache cacheable responses (like browsers do).
//
// When an http request is made, a peer is chosen to handle the request
// according to the requested url's canonical owner.
//
// If the content is cacheable as per the HTTP RFC, it will get cached
// on the peer and the response is then returned to the client. (thanks
// to github.com/gregjones/httpcache)
//
// Note that the peers are not real HTTP proxies. They are themselves
// querying the origin servers and copying the response back to clients.
// It has the benefit of being able to cache TLS requests.
//
// The result is that it is only suitable for use as a distributed 'private'
// cache since the requests are fully intercepted by the peers.
package forwardcache

import (
	"net/http"
	"net/http/httputil"

	"github.com/gregjones/httpcache"
)

// Peer is a peer in the pool. It handles and cache the requests for the clients.
// It is also able to issue requests on its own to other peers when a resource
// belongs to it.
type Peer struct {
	*Client
	handler   *proxy
	self      string
	cache     httpcache.Cache
	transport http.RoundTripper
	buffers   httputil.BufferPool
}

// NewPeer creates a Peer.
// The returned *Peer implements http.Handler and must be registered
// manually using http.Handle to serve local requests. See Handler().
func NewPeer(self string, options ...func(*Peer)) *Peer {
	p := &Peer{
		Client:    NewClient(),
		self:      self,
		transport: http.DefaultTransport,
		cache:     httpcache.NewMemoryCache(),
	}

	for _, option := range options {
		option(p)
	}

	p.handler = newProxy(p.Client.path, p.cache, p.transport, p.buffers)
	return p
}

// Handler returns an http.Handler to be registered using http.Handle
// for the local Peer to serve requests.
func (p *Peer) Handler() http.Handler {
	return p.handler
}

// RoundTrip makes the request go through one of the peer using its internal
// Client. If the local peer is targeted, it uses the local handler directly.
// Since Peer implements the Roundtripper interface, it can be used as a transport.
func (p *Peer) RoundTrip(req *http.Request) (*http.Response, error) {
	peer := p.Client.choosePeer(req.URL.String())

	if peer == p.self {
		return p.handler.Transport.RoundTrip(req)
	}

	return p.Client.roundTripTo(peer, req)
}

// WithClient lets you configure a custom pool client.
// Defaults to NewClient(). If a Client is not specified
// upon Peer creation, SetPool(...) must be called to set
// the list of peers.
func WithClient(c *Client) func(*Peer) {
	return func(p *Peer) {
		p.Client = c
	}
}

// WithPeerTransport lets you configure a custom
// transport used between the local peer and origins.
// Defaults to http.DefaultTransport.
func WithPeerTransport(t http.RoundTripper) func(*Peer) {
	return func(p *Peer) {
		p.transport = t
	}
}

// WithBufferPool lets you configure a custom buffer pool.
// Defaults to not using a buffer pool.
func WithBufferPool(b httputil.BufferPool) func(*Peer) {
	return func(p *Peer) {
		p.buffers = b
	}
}

// WithDefaultBufferPool lets you use the default 32k buffer pool.
// Defaults to not using a buffer pool.
func WithDefaultBufferPool(b httputil.BufferPool) func(*Peer) {
	return func(p *Peer) {
		p.buffers = DefaultBufferPool
	}
}

// WithCache lets you use a custom httpcache.Cache.
// Defaults to httpcache.MemoryCache.
func WithCache(c httpcache.Cache) func(*Peer) {
	return func(p *Peer) {
		p.cache = c
	}
}


proxy.go
/*
Copyright 2018 Mike Gleason jr Couturier.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package forwardcache

import (
	"context"
	"net/http"
	"net/http/httputil"
	"net/url"

	"github.com/gregjones/httpcache"
)

type key int

const originKey key = 1

// proxy is the forward caching proxy on a peer, it uses
// a cache that conforms to the HTTP RFC (thanks to
// github.com/gregjones/httpcache)
type proxy struct {
	path string
	*httputil.ReverseProxy
}

// newProxy creates a proxy that serves requests on path using the
// specified cache. The proxy handles requests of format:
// /path?q=originUrl where originUrl is the resource being
// requested by the client.
func newProxy(path string, cache httpcache.Cache, transport http.RoundTripper, buffers httputil.BufferPool) *proxy {
	return &proxy{
		path: path,
		ReverseProxy: &httputil.ReverseProxy{
			Transport: &httpcache.Transport{
				Cache:               cache,
				MarkCachedResponses: true,
				Transport:           transport,
			},
			Director:   director,
			BufferPool: buffers,
		},
	}
}

// ServeHTTP takes the url of the requested resource to be fetched on the
// origin and puts in in the request's context to be used later by the proxy director.
func (p *proxy) ServeHTTP(w http.ResponseWriter, req *http.Request) {
	if req.URL.Path != p.path {
		w.WriteHeader(http.StatusBadGateway)
		return
	}

	q := req.URL.Query().Get("q")
	if q == "" {
		w.WriteHeader(http.StatusBadGateway)
		return
	}

	origin, err := url.Parse(q)
	if err != nil {
		w.WriteHeader(http.StatusBadGateway)
		return
	}

	ctx := context.WithValue(req.Context(), originKey, origin)
	p.ReverseProxy.ServeHTTP(w, req.WithContext(ctx))
}

// director modifies the requested URL to the origin.
func director(req *http.Request) {
	origin := req.Context().Value(originKey).(*url.URL)
	req.URL = origin
	req.Host = origin.Host
}


client.go
package forwardcache

import (
	"hash/crc32"
	"net/http"
	"net/url"
	"sync"

	"github.com/mikegleasonjr/forwardcache/consistenthash"
)

const (
	defaultPath     = "/proxy"
	defaultReplicas = 50
)

// Client represents a nonparticipating client in the pool. It delegates
// requests to the responsible peer.
type Client struct {
	path      string
	replicas  int
	hashFn    consistenthash.Hash
	transport http.RoundTripper
	peers     []string
	mu        sync.RWMutex // guards peers
	hashMap   *consistenthash.Map
}

// NewClient creates a Client.
func NewClient(options ...func(*Client)) *Client {
	c := &Client{
		path:      defaultPath,
		replicas:  defaultReplicas,
		hashFn:    crc32.ChecksumIEEE,
		transport: http.DefaultTransport,
	}

	for _, option := range options {
		option(c)
	}

	c.SetPool(c.peers...)
	return c
}

// SetPool updates the client's peers list. Each peer should
// be a valid base URL, for example "http://example.net:8000".
func (c *Client) SetPool(peers ...string) {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.peers = peers
	c.hashMap = consistenthash.New(c.replicas, c.hashFn)
	c.hashMap.Add(c.peers...)
}

// HTTPClient returns an http.Client that uses the Client as its transport.
func (c *Client) HTTPClient() *http.Client {
	cl := new(http.Client)
	*cl = *http.DefaultClient
	cl.Transport = c
	return cl
}

// RoundTrip makes the request go through one of the peer. Since Client
// implements the Roundtripper interface, it can be used as a transport.
func (c *Client) RoundTrip(req *http.Request) (*http.Response, error) {
	peer := c.choosePeer(req.URL.String())
	return c.roundTripTo(peer, req)
}

func (c *Client) choosePeer(url string) string {
	c.mu.RLock()
	defer c.mu.RUnlock()

	return c.hashMap.Get(url)
}

func (c *Client) roundTripTo(peer string, req *http.Request) (*http.Response, error) {
	query := c.peerHandlerURL(peer, req.URL.String())

	cpy := clone(req) // per RoundTripper contract
	cpy.URL = query
	cpy.Host = query.Host

	return c.transport.RoundTrip(cpy)
}

func (c *Client) peerHandlerURL(peer string, origin string) *url.URL {
	u, _ := url.Parse(peer)

	u.Path = c.path
	u.RawQuery = "q=" + url.QueryEscape(origin)

	return u
}

// WithPath specifies the HTTP path that will serve proxy requests.
// Defaults to "/proxy".
func WithPath(p string) func(*Client) {
	return func(c *Client) {
		c.path = p
	}
}

// WithReplicas specifies the number of key replicas on the consistent hash.
// Defaults to 50.
func WithReplicas(r int) func(*Client) {
	return func(c *Client) {
		c.replicas = r
	}
}

// WithHashFn specifies the hash function of the consistent hash.
// Defaults to crc32.ChecksumIEEE.
func WithHashFn(h consistenthash.Hash) func(*Client) {
	return func(c *Client) {
		c.hashFn = h
	}
}

// WithClientTransport lets you configure a custom transport
// used between the local client and the proxies.
// Defaults to http.DefaultTransport.
func WithClientTransport(t http.RoundTripper) func(*Client) {
	return func(c *Client) {
		c.transport = t
	}
}

// WithPool lets you configure the client's list of peers.
// Defaults to nil. See Client.SetPool(...).
func WithPool(peers ...string) func(*Client) {
	return func(c *Client) {
		c.peers = peers
	}
}

// clones a request, credits goes to:
// https://github.com/golang/oauth2/blob/master/transport.go#L36
func clone(r *http.Request) *http.Request {
	r2 := new(http.Request)
	*r2 = *r
	r2.Header = make(http.Header, len(r.Header))
	for k, s := range r.Header {
		r2.Header[k] = append([]string(nil), s...)
	}
	return r2
}


forwardcache.go
/*
Copyright 2018 Mike Gleason jr Couturier.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package forwardcache provides a forward caching proxy that works
// across a set of peer processes.
//
// When an http request is made, a peer is chosen to proxy the request
// according to the requested url's canonical owner.
//
// If the content is cacheable as per the HTTP RFC, it will get cached
// on the proxy and the response is then returned to the client. (thanks
// to github.com/gregjones/httpcache)
//
// Note that the peers are not real HTTP proxies. They are themselves
// querying the origin servers and copying the response back to clients.
// It has the benefit of being able to cache TLS requests.
//
// The result is that it is only suitable for use as a distributed 'private'
// cache since the requests are fully intercepted by the peers.
package forwardcache

import (
	"hash/crc32"
	"net/http"
	"net/http/httputil"
	"net/url"
	"sync"

	"github.com/gregjones/httpcache"
	"github.com/mikegleasonjr/forwardcache/consistenthash"
)

const (
	defaultPath     = "/proxy"
	defaultReplicas = 50
)

// Pool represents all caching proxies spread over 1 or more machines. It
// also acts as a participating peer.
type Pool struct {
	*Client
	self      string
	local     *proxy
	transport http.RoundTripper
	buffers   httputil.BufferPool
}

// NewPool creates a Pool and registers itself using the specified cache.
// The returned *Pool implements http.Handler and must be registered
// manually using http.Handle to serve the local proxy. See LocalProxy()
func NewPool(self string, local httpcache.Cache, options ...func(*Pool)) *Pool {
	p := &Pool{self: self}
	for _, option := range options {
		option(p)
	}
	if p.transport == nil {
		p.transport = http.DefaultTransport
	}
	if p.Client == nil {
		p.Client = NewClient()
	}
	p.local = newProxy(p.path, local, p.transport, p.buffers)
	return p
}

// LocalProxy returns an http.Handler to be registered using http.Handle
// for the local proxy to serve requests.
func (p *Pool) LocalProxy() http.Handler {
	return p.local
}

// WithProxyTransport lets you configure a custom
// transport used between the local proxy and the origins.
// Defaults to http.DefaultTransport.
func WithProxyTransport(t http.RoundTripper) func(*Pool) {
	return func(p *Pool) {
		p.transport = t
	}
}

// WithClient lets you configure a custom pool client.
// Defaults to NewClient().
func WithClient(c *Client) func(*Pool) {
	return func(p *Pool) {
		p.Client = c
	}
}

// WithBufferPool lets you configure a custom buffer pool.
// Defaults to not using a buffer pool.
func WithBufferPool(b httputil.BufferPool) func(*Pool) {
	return func(p *Pool) {
		p.buffers = b
	}
}

// WithDefaultBufferPool lets you use the default 32k buffer pool.
// Defaults to not using a buffer pool.
func WithDefaultBufferPool(b httputil.BufferPool) func(*Pool) {
	return func(p *Pool) {
		p.buffers = DefaultBufferPool
	}
}

// Client represents a nonparticipating client in the pool. It can
// issue requests to the pool but not proxy requests for others.
type Client struct {
	path      string
	replicas  int
	hashFn    consistenthash.Hash
	transport http.RoundTripper
	mu        sync.RWMutex // guards peers
	peers     *consistenthash.Map
}

// NewClient creates a Client.
func NewClient(options ...func(*Client)) *Client {
	c := &Client{
		path:      defaultPath,
		replicas:  defaultReplicas,
		hashFn:    crc32.ChecksumIEEE,
		transport: http.DefaultTransport,
	}
	for _, option := range options {
		option(c)
	}
	return c
}

// Set updates the pool's list of peers. Each peer value should
// be a valid base URL, for example "http://example.net:8000".
func (c *Client) Set(peers ...string) {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.peers = consistenthash.New(c.replicas, c.hashFn)
	c.peers.Add(peers...)
}

// HTTPClient returns an http.Client that uses the pool as its transport.
func (c *Client) HTTPClient() *http.Client {
	cl := new(http.Client)
	*cl = *http.DefaultClient
	cl.Transport = c
	return cl
}

// RoundTrip makes the request go through one of the proxy. If the local
// proxy is targetted, it uses the local transport directly. Since Client
// implements the Roundtripper interface, it can be used as a transport.
func (c *Client) RoundTrip(req *http.Request) (*http.Response, error) {
	c.mu.RLock()
	peer := c.peers.Get(req.URL.String())
	c.mu.RUnlock()

	cpy := clone(req) // per RoundTripper contract
	query := proxyHandlerURL(peer, c.path, cpy.URL.String())
	cpy.URL = query
	cpy.Host = query.Host

	return c.transport.RoundTrip(cpy)
}

// WithPath specifies the HTTP path that will serve proxy requests.
// Defaults to "/proxy".
func WithPath(p string) func(*Client) {
	return func(c *Client) {
		c.path = p
	}
}

// WithReplicas specifies the number of key replicas on the consistent hash.
// Defaults to 50.
func WithReplicas(r int) func(*Client) {
	return func(c *Client) {
		c.replicas = r
	}
}

// WithHashFn specifies the hash function of the consistent hash.
// Defaults to crc32.ChecksumIEEE.
func WithHashFn(h consistenthash.Hash) func(*Client) {
	return func(c *Client) {
		c.hashFn = h
	}
}

// WithClientTransport lets you configure a custom transport
// used between the local client and the proxies.
// Defaults to http.DefaultTransport.
func WithClientTransport(t http.RoundTripper) func(*Client) {
	return func(c *Client) {
		c.transport = t
	}
}

// builds the url that handles proxy requests on the selected peer
func proxyHandlerURL(peer, path, origin string) *url.URL {
	u, _ := url.Parse(peer)
	u.Path = path
	u.RawQuery = "q=" + url.QueryEscape(origin)
	return u
}

// clones a request, credits goes to:
// https://github.com/golang/oauth2/blob/master/transport.go#L36
func clone(r *http.Request) *http.Request {
	r2 := new(http.Request)
	*r2 = *r
	r2.Header = make(http.Header, len(r.Header))
	for k, s := range r.Header {
		r2.Header[k] = append([]string(nil), s...)
	}
	return r2
}
